{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d294621-eae9-4aa7-9c74-ad36cc46c3d3",
   "metadata": {},
   "source": [
    "# 10X Visium HD Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b4f0b4-c8d6-4140-92c0-c2b1f7a2fc14",
   "metadata": {},
   "source": [
    "rakaia supports the visualization and anlaysis of spot and bin-based spatial transcriptomic technologies such as 10X Visium. The preprocessing steps required for Visium HD datasets differ slightly from 10X Visium V1 and V2 data, as the datasets are binned and exported at different micron resolutions. Here we demonstrate how to extract different micron resolution files for visualization in rakaia. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bbb03939-248f-402d-8dde-b3f7cef80286",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spatialdata_io import visium_hd\n",
    "import spatialdata as sd\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "input_dir = \"/home/admin/rakaia/visium/hd/\"\n",
    "zarr_path = os.path.join(input_dir, 'zarr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04dca5af-ba73-413b-8edb-4dc60e516980",
   "metadata": {},
   "source": [
    "We will use the mouse small intestine Visium HD dataset that is publicly available from 10X Genomics [here](https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-intestine). Using the link provided, select **Batch download**, and download all of the files shown to the desired input directory. Once the download has complete, the *tar.gz* should be extracted into the input directory, producing a sub-directory named **binned_outputs**. \n",
    "\n",
    "Next, you should specify the experiment name variable below by checking the anallysis summary HTML or related output files. For example, the summary [here](https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_HD_Mouse_Small_Intestine/Visium_HD_Mouse_Small_Intestine_web_summary.html) shows the dataset name in the title. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "47c15892-f406-4d13-bb21-acbe7b7331ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_name = \"Visium_HD_Mouse_Small_Intestine\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46e95c05-bbac-4aee-ad64-a6a9f55fe2c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SpatialData object, with associated Zarr store: /home/admin/rakaia/visium/hd/zarr\n",
       "├── Images\n",
       "│     ├── 'Visium_HD_Mouse_Small_Intestine_hires_image': DataArray[cyx] (3, 5575, 6000)\n",
       "│     └── 'Visium_HD_Mouse_Small_Intestine_lowres_image': DataArray[cyx] (3, 558, 600)\n",
       "├── Shapes\n",
       "│     ├── 'Visium_HD_Mouse_Small_Intestine_square_002um': GeoDataFrame shape: (5479660, 1) (2D shapes)\n",
       "│     ├── 'Visium_HD_Mouse_Small_Intestine_square_008um': GeoDataFrame shape: (351817, 1) (2D shapes)\n",
       "│     └── 'Visium_HD_Mouse_Small_Intestine_square_016um': GeoDataFrame shape: (91033, 1) (2D shapes)\n",
       "└── Tables\n",
       "      ├── 'square_002um': AnnData (5479660, 19059)\n",
       "      ├── 'square_008um': AnnData (351817, 19059)\n",
       "      └── 'square_016um': AnnData (91033, 19059)\n",
       "with coordinate systems:\n",
       "    ▸ 'downscaled_hires', with elements:\n",
       "        Visium_HD_Mouse_Small_Intestine_hires_image (Images), Visium_HD_Mouse_Small_Intestine_square_002um (Shapes), Visium_HD_Mouse_Small_Intestine_square_008um (Shapes), Visium_HD_Mouse_Small_Intestine_square_016um (Shapes)\n",
       "    ▸ 'downscaled_lowres', with elements:\n",
       "        Visium_HD_Mouse_Small_Intestine_lowres_image (Images), Visium_HD_Mouse_Small_Intestine_square_002um (Shapes), Visium_HD_Mouse_Small_Intestine_square_008um (Shapes), Visium_HD_Mouse_Small_Intestine_square_016um (Shapes)\n",
       "    ▸ 'global', with elements:\n",
       "        Visium_HD_Mouse_Small_Intestine_square_002um (Shapes), Visium_HD_Mouse_Small_Intestine_square_008um (Shapes), Visium_HD_Mouse_Small_Intestine_square_016um (Shapes)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    sdata = visium_hd(input_dir)\n",
    "    sdata.write(zarr_path)\n",
    "except ValueError:\n",
    "    pass\n",
    "sdata = sd.read_zarr(zarr_path)\n",
    "sdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a40564f-00ff-4dd8-a016-3552c2591901",
   "metadata": {},
   "source": [
    "As seen above, the Visium HD spatialdata object contains a variety of images, shapes, and coordinate systems. For rakaia visualization, the Anndata tables will be scaled and exported. \n",
    "\n",
    "By default, Visium HD datasets summarize expression levels at three different micron \"bin\" resolutions: 2, 8, and 16 microns. Smaller bin sizes produce higher resolution datasets that are closer to the original image resolution, but are slower to visualize in rakaia. It is recommended to use either 8 or 16 micron resolutions in rakaia for performance purposes. \n",
    "\n",
    "Below, we will extract the feature objects needed for visualization and scale the spatial coordinates by the micron resolution. This is similar to the procedure described in spatialdata [here](https://spatialdata.scverse.org/en/latest/tutorials/notebooks/notebooks/examples/technlology_visium_hd.html#performant-on-the-fly-data-rasterization), where rasterizing the expressoin for each bin level produces an image frame that can be queried. We will do this inside the Anndata object to produce an output whose size and resolution is scaled to the bin resolution. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6151b29a-2a2c-4205-8c2d-2128ae1bb955",
   "metadata": {},
   "outputs": [],
   "source": [
    "for bin_size in [2, 8, 16]:\n",
    "    extra_zero = \"0\" if bin_size < 16 else \"\"\n",
    "    adata = sdata.tables[f\"square_0{extra_zero}{bin_size}um\"]\n",
    "    # sdata.tables[f\"square_0{extra_zero}{bin_size}um\"].X = (\n",
    "    #     sdata.tables[f\"square_0{extra_zero}{bin_size}um\"].X.tocsc())\n",
    "    # rasterized = sd.rasterize_bins(\n",
    "    #             sdata,\n",
    "    #             f\"{experiment_name}_square_0{extra_zero}{bin_size}um\",\n",
    "    #             f\"square_0{extra_zero}{bin_size}um\",\n",
    "    #             \"array_col\",\n",
    "    #             \"array_row\",\n",
    "    #         )\n",
    "    # # for now, don't use the rasterized xarray frames as they are too large to cast to sparse numpy\n",
    "    # sdata[f\"rasterized_{bin_size}um\"] = rasterized\n",
    "    adata.obsm['spatial'] = adata.obsm['spatial'] / float(bin_size)\n",
    "    adata.var_names_make_unique()\n",
    "    adata.uns[\"scaling_visium_hd\"] = int(bin_size)\n",
    "    adata.write_h5ad(os.path.join(input_dir, \"anndata\", f\"{dataset_name}_{bin_size}um.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12253d8c-88f4-4e27-8431-44a97adb3c3d",
   "metadata": {},
   "source": [
    "**IMPORTANT**: The `scaling_visium_hd` key in the uns slot is required to align the h5ad expression values to whole slide images (WSI) such as H & E in rakaia. \n",
    "\n",
    "The outputs are three h5ad files containing the scaled expression for each bin level as separate input files. When imported into rakaia, it can be seen that each file has different dimensions, with the largest bin size (16) generating the smallest dimensions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5843b805-609c-48ac-8511-56424d704648",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
