{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "023c4241-51da-4179-b6e0-563fe16c77db",
   "metadata": {},
   "source": [
    "# Xenium preprocessing\n",
    "\n",
    "rakaia provides the ability to read and visualize spatial transcriptomic datasets such as Visium and Xenium. For Xenium, rakaia supports the overlay of segmentation masks alongside aggregated transcript counts per cell. To be consistent with the format of other technologies, the transcript counts and segmentation mask need to be imported as separate Anndata and tiff objects, respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c0cd02c-70e4-4cc6-9f7c-317008f3aa7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import spatialdata as sd\n",
    "from spatialdata_io import xenium\n",
    "import rasterio\n",
    "from rasterio.features import rasterize\n",
    "import numpy as np\n",
    "import tifffile as tiff\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# Set the input and zarr output directories to whatever you want\n",
    "xenium_path = \"/home/admin/github/rakaia/sandbox/Xenium/\"\n",
    "zarr_path = \"/home/admin/github/rakaia/sandbox/Xenium.zarr/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e61cee2-e996-4685-9139-27dfd6048f84",
   "metadata": {},
   "source": [
    "We will use the Xenium human melanoma dataset from [here](https://www.10xgenomics.com/datasets/human-skin-preview-data-xenium-human-skin-gene-expression-panel-add-on-1-standard) as an example. Download the **Xenium Output bundle (full)** from the Output and supplemental files section in the link, and extract/unzip the contents to the same destination directory as *xenium_path* above.\n",
    "\n",
    "We will next read the data into a SpatialData object and pull the transcript information with spatial coordinates as an Anndata object. It is important to note that in the Anndata object, the transcript counts are summarized per cell, and the coordinates provided are the approximate centroid of each cell. This may differ from the 10X Xenium explorer, which may render individual transcripts in the same cell with slightly different coordinates/ z-level depending on the resolution zoom level. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6fd5cec1-6de8-4d21-ac1a-2578251f1795",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m reading \u001b[35m/home/admin/github/rakaia/sandbox/Xenium/\u001b[0m\u001b[95mcell_feature_matrix.h5\u001b[0m                                   \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 87499 × 382\n",
       "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'spatialdata_attrs'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    sdata = xenium(xenium_path)\n",
    "    sdata.write(zarr_path)\n",
    "# if the directory already exists, can pass this step\n",
    "except ValueError:\n",
    "    pass\n",
    "\n",
    "sdata = sd.read_zarr(zarr_path)\n",
    "\n",
    "# get the transcripts in Anndata format from the table slot\n",
    "adata = sdata['table']\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7620ae56-88e9-40ca-8086-938c0248f886",
   "metadata": {},
   "source": [
    "Next, we will extract the cell boundaries for the matching segmentation mask, and infer the coordinate bounds to convert the segmentation polygons into a segmentation array. These bounds will match the transcript arrays to the segmentation mask in rakaia. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a2a02606-881b-45ec-9ce4-c9307a0ab0b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the cell segmentation data frame\n",
    "cells = sdata.shapes['cell_boundaries']\n",
    "\n",
    "# get the int bounds to know where the segmentation mask is\n",
    "x_min, y_min = np.min((adata.obsm['spatial']), axis=0)\n",
    "x_max, y_max = np.max((adata.obsm['spatial']), axis=0)\n",
    "\n",
    "\n",
    "transform = rasterio.transform.from_bounds(x_min, y_min, x_max,\n",
    "                                           y_max, int(x_max - x_min), int(y_max - y_min))\n",
    "\n",
    "shapes = [(geom, idx) for idx, geom in enumerate(cells.geometry)]\n",
    "\n",
    "# Set the mask shape to be the same as the transcript bounds in rakaia\n",
    "mask = rasterize(shapes=shapes, out_shape=(int(y_max - y_min), int(x_max - x_min)),\n",
    "                 transform=transform, dtype='int32')\n",
    "\n",
    "# Set the name of the output directory and give the outputs a relevant name\n",
    "outdir = \"/home/admin/rakaia/xenium/\"\n",
    "out_name = \"melanoma\"\n",
    "\n",
    "\n",
    "# NOTE: in most instances, the coordinates along the y axis \n",
    "# need to be flipped prior to export using np.flip with axis 0\n",
    "tiff.imwrite(f\"{outdir}/{out_name}_mask.tiff\", np.flip(mask, axis=0))\n",
    "adata.write_h5ad(f\"{outdir}/{out_name}.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e59a60-05ea-48a6-b67f-3ff7f546e298",
   "metadata": {},
   "source": [
    "The resulting output files can be found in the specified **outdir** and can be imported into rakaia using the corresponding uploader components. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e12276-e788-4daa-b36a-974f24be1387",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
